tests/lib/rules/no-this-in-sfc.js

```diff
@@ -112,16 +112,6 @@ ruleTester.run('no-this-in-sfc', rule, {
         };
       }
     }`
-  }, {
-    code: `
-    class Foo {
-      bar() {
-        () => () => {
-          this.something();
-          return null;
-        };
-      }
```

Comment:
hmm, these aren't components, and the `this` here refers to the class instance. Why are these invalid?

Reply:
@ljharb 
In this test file there is the following example:

https://github.com/yannickcr/eslint-plugin-react/blob/4c8d8cc0a5dca39554cf97b807822b60782f320f/tests/lib/rules/no-this-in-sfc.js#L222

```js
{
    code: `
    class Foo {
      bar() {
        function Bar(){
          return () => {
            this.something();
            return null;
          }
        }
      }
    }`,
    errors: [{message: ERROR_MESSAGE}]
  }
```

You can also try this one by removing the name of the function:

```js
{
    code: `
    class Foo {
      bar() {
        return function () {
          return () => {
            this.something();
            return null;
          }
        }
      }
    }`,
    errors: [{message: ERROR_MESSAGE}]
  }
```

The purpose of this PR is to make the behavior consistent. If the two examples using `function` keyword are invalid why is the example using `arrow function` valid? In the two examples I showed the arrow functions are also using the `this` coming from the outer function and the outer functions are not a React components since they don't return null or JSX.

Reply:
I think that any linting warning for this rule that's produced on "not a React component" is a bug. A class that doesn't extend `React.Component` (or have the appropriate jsdoc) should be ignored (altho if it contains a react component, that should be considered)

Reply:
> I think that any linting warning for this rule that's produced on "not a React component" is a bug. A class that doesn't extend `React.Component` (or have the appropriate jsdoc) should be ignored (altho if it contains a react component, that should be considered)

Yeah, agree but the error is not on the `class Foo` but on the returned arrow function.

I suppose that in other to support HOC, functions that return jsx or null and, are returned by other function (using the `return` keyword) are considered components. For example:

```js
function connect(Component) {
  return () => <Component>{this.props.one}</Component> // considered a component
}

const connect = (Component) => {
  return () => <Component>{this.props.one}</Component> // considered a component
}

const connect = (Component) => {
  return () => { // considered a component
    return null;
  }
}
```

What this PR is trying to fix is that their counter part arrow functions (without `return` keyword) are not considered components:

```js
const connect = (Component) => (
  () => <Component>{this.props.one}</Component>  // not considered a component
)

const connect = (Component) => (
  () => { // not considered a component
    return null;
  }
)
```

Reply:
hmm, ok - but an arrow function that returns an arrow function that returns null doesn't seem likely to be a component to me.

Reply:
> hmm, ok - but an arrow function that returns an arrow function that returns null doesn't seem likely to be a component to me.

Totally agree but, supporting HOC and functions returning null considered components makes to correctly identify components in this situation incredible hard. Maybe we could enforce in this situation the need for returning JSX and not just null.

Following your example about *arrow function that returns arrow function that returns null*, this is a component right now:

```js
const connect = () => {
  return () => {
    return () => { // <-- a component
      return null;
    }
  };
}
```
But, the inconsistent behavior, this is not:

```js
const connect = () => {
  return () => (
    () => { // <-- not a component
      return null;
    }
  );
}
```